utils/s3/client.cc

```diff
@@ -98,6 +98,191 @@ future<> client::delete_object(sstring object_name) {
     co_await _http.make_request(std::move(req), ignore_reply, http::reply::status_type::no_content);
 }
 
+class client::upload_sink : public data_sink_impl {
+    // "Each part must be at least 5 MB in size, except the last part."
+    // https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPart.html
+    static constexpr size_t minimum_part_size = 5 << 20;
+    static constexpr int flush_concurrency = 3;
+
+    shared_ptr<client> _client;
+    http::experimental::client& _http;
+    sstring _object_name;
+    memory_data_sink_buffers _bufs;
+    sstring _upload_id;
+    utils::chunked_vector<sstring> _part_etags;
+    semaphore _flush_sem{flush_concurrency};
+
+    future<> start_upload();
+    future<> finalize_upload();
+    future<> maybe_flush();
+    future<> do_flush();
+    future<> upload_part(unsigned part_number, memory_data_sink_buffers bufs);
+    future<> abort_upload();
+
+    bool upload_started() const noexcept {
+        return !_upload_id.empty();
+    }
+
+public:
+    upload_sink(shared_ptr<client> cln, sstring object_name)
+        : _client(std::move(cln))
+        , _http(_client->_http)
+        , _object_name(std::move(object_name))
+    {
+    }
+
+    virtual future<> put(net::packet) override {
+        throw_with_backtrace<std::runtime_error>("s3 put(net::packet) unsupported");
+    }
+
+    virtual future<> put(temporary_buffer<char> buf) override {
+        _bufs.put(std::move(buf));
+        return maybe_flush();
+    }
+
+    virtual future<> put(std::vector<temporary_buffer<char>> data) override {
+        for (auto&& buf : data) {
+            _bufs.put(std::move(buf));
+        }
+        return maybe_flush();
+    }
+
+    virtual future<> flush() override {
+        return finalize_upload();
+    }
+
+    virtual future<> close() override;
+
+    virtual size_t buffer_size() const noexcept override {
+        return 128 * 1024;
+    }
+};
+
+future<> client::upload_sink::maybe_flush() {
+    if (_bufs.size() >= minimum_part_size) {
+        co_await do_flush();
+    }
+}
+
+future<> client::upload_sink::do_flush() {
+    if (!upload_started()) {
+        co_await start_upload();
+    }
+    auto pn = _part_etags.size();
+    _part_etags.emplace_back();
+    co_await upload_part(pn, std::move(_bufs));
+}
+
+sstring parse_multipart_upload_id(sstring& body) {
+    rapidxml::xml_document<> doc;
+    std::exception_ptr ex;
+    try {
+        doc.parse<0>(body.data());
+    } catch (const rapidxml::parse_error& e) {
+        s3l.warn("cannot parse initiate multipart upload response: {}", e.what());
+        return "";
```

Comment:
Shouldn't we throw? Or can we recover?

Reply:
The caller checks for the returned value to be empty and resolves with relevant exceptional future

Reply:
Ok. Please document it whereever we swallow an exception, it screams "there is an error handling bug here".